File: problems/0008_string-to-integer-atoi/built-in/Solution.py
#!/usr/bin/python3

import re

class Solution:
    def myAtoi(self, str: str) -> int:
        matcher = Solution.NUM_FMT.match(str)
        if matcher is not None:
            try:
                result = int(matcher.group(1))
                if result < Solution.INT_MIN:
                    return Solution.INT_MIN
                elif result > Solution.INT_MAX:
                    return Solution.INT_MAX
                else:
                    return result
            except ValueError:
                pass
        return 0
    NUM_FMT = re.compile(r'\s*([\-\+]?\d+).*')
    INT_MIN = -2**31
    INT_MAX = 2**31 - 1

File: problems/0008_string-to-integer-atoi/built-in/Solution.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.math.BigInteger;

class Solution {
    public int myAtoi(String str) {
        try {
            final Matcher matcher = NUM_FMT.matcher(str);
            if (matcher.matches()) {
                final BigInteger result = new BigInteger(matcher.group(1));
                if (result.compareTo(INT_MIN) < 0)
                    return Integer.MIN_VALUE;
                else if (result.compareTo(INT_MAX) > 0)
                    return Integer.MAX_VALUE;
                else
                    return result.intValue();
            } else
                return 0;
        } catch (final NumberFormatException ignored) {
            return 0;
        }
    }
    static final Pattern NUM_FMT = Pattern.compile("\\s*([\\-\\+]?\\d+).*");
    static final BigInteger INT_MIN = BigInteger.valueOf(Integer.MIN_VALUE),
                            INT_MAX = BigInteger.valueOf(Integer.MAX_VALUE);
}
